changeset:   297:3ac205ad1f05
user:        mcimadamore
date:        Tue Jun 16 10:46:16 2009 +0100
summary:     6835428: regression: return-type inference rejects valid code

diff -r 3d539f4123b8 -r 3ac205ad1f05 src/share/classes/com/sun/tools/javac/comp/Infer.java
--- a/src/share/classes/com/sun/tools/javac/comp/Infer.java	Tue Jun 16 10:45:42 2009 +0100
+++ b/src/share/classes/com/sun/tools/javac/comp/Infer.java	Tue Jun 16 10:46:16 2009 +0100
@@ -158,8 +158,7 @@
                 that.inst = types.glb(that.hibounds);
         }
         if (that.inst == null ||
-            that.inst.isErroneous() ||
-            !types.isSubtypeUnchecked(that.inst, that.hibounds, warn))
+            that.inst.isErroneous())
             throw ambiguousNoInstanceException
                 .setMessage("no.unique.maximal.instance.exists",
                             that.qtype, that.hibounds);
diff -r 3d539f4123b8 -r 3ac205ad1f05 test/tools/javac/generics/inference/T6835428.java
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/test/tools/javac/generics/inference/T6835428.java	Tue Jun 16 10:46:16 2009 +0100
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2009 Sun Microsystems, Inc.  All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara,
+ * CA 95054 USA or visit www.sun.com if you need additional information or
+ * have any questions.
+ */
+
+/**
+ * @test
+ * @bug 6835428
+ * @author mcimadamore
+ * @summary regression: return-type inference rejects valid code
+ * @compile T6835428.java
+ */
+
+class T6835428<T> {
+    interface X<T> {}
+   <T extends Comparable<? super T>> T6835428<X<T>> m() { return null; }
+   <T extends Comparable<? super T>> void test() {
+      T6835428<X<T>> t = m();
+   }
+}

